flink-runtime/src/main/java/org/apache/flink/runtime/codegeneration/SorterFactory.java

```diff
@@ -81,6 +81,7 @@
 	 */
 	private SorterFactory() {
 		this.classCompiler = new SimpleCompiler();
+		this.classCompiler.setParentClassLoader(this.getClass().getClassLoader());
```

Comment:
I think that the `SorterFactory` class is not instantiated by the user code classloader, so `this.getClass().getClassLoader()` gets the wrong classloader here.

Instead, we should get the user code classloader from the caller of `createSorter`.

I'll make this change, and also do the modifications to the cache, as discussed above.